exp/internal/controllers/machinepool_controller_noderef.go

```diff
@@ -221,6 +202,39 @@ func (r *MachinePoolReconciler) getNodeReferences(ctx context.Context, c client.
 	return getNodeReferencesResult{nodeRefs, available, ready}, nil
 }
 
+// patchNodes patches the nodes with the cluster name and cluster namespace annotations.
+func (r *MachinePoolReconciler) patchNodes(ctx context.Context, c client.Client, references []corev1.ObjectReference, mp *expv1.MachinePool) error {
+	log := ctrl.LoggerFrom(ctx)
+	for _, nodeRef := range references {
+		node := &corev1.Node{}
+		if err := c.Get(ctx, client.ObjectKey{Name: nodeRef.Name}, node); err != nil {
+			log.V(2).Info("Failed to get Node, skipping setting annotations", "err", err, "nodeRef.Name", nodeRef.Name)
+			continue
+		}
+		patchHelper, err := patch.NewHelper(node, c)
+		if err != nil {
+			return err
+		}
+		desired := map[string]string{
+			clusterv1.ClusterNameAnnotation:      mp.Spec.ClusterName,
+			clusterv1.ClusterNamespaceAnnotation: mp.GetNamespace(),
+			clusterv1.OwnerKindAnnotation:        mp.Kind,
+			clusterv1.OwnerNameAnnotation:        mp.Name,
+		}
+		// Add annotations and drop NodeUninitializedTaint.
```

Comment:
this is the real change here. Everything else is just extracting the code into a function so it can be unit tested.